geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -3929,27 +3957,19 @@ public void setGatewayConflictResolver(GatewayConflictResolver resolver) {
 
   @Override
   public List<CacheServer> getCacheServers() {
-    List<CacheServer> cacheServersWithoutReceiver = null;
-    if (!this.allCacheServers.isEmpty()) {
-      for (CacheServerImpl cacheServer : this.allCacheServers) {
-        // If CacheServer is a GatewayReceiver, don't return as part of CacheServers
-        if (!cacheServer.isGatewayReceiver()) {
-          if (cacheServersWithoutReceiver == null) {
-            cacheServersWithoutReceiver = new ArrayList<>();
-          }
-          cacheServersWithoutReceiver.add(cacheServer);
-        }
-      }
-    }
-    if (cacheServersWithoutReceiver == null) {
-      cacheServersWithoutReceiver = Collections.emptyList();
-    }
-    return cacheServersWithoutReceiver;
+    return Collections.unmodifiableList(allCacheServers);
   }
 
   @Override
-  public List getCacheServersAndGatewayReceiver() {
-    return this.allCacheServers;
+  public List<InternalCacheServer> getCacheServersAndGatewayReceiver() {
```

Comment:
You may have competing changes to https://github.com/apache/geode/pull/3510. Please check with @dschneider-pivotal for performance issues in this area.

Reply:
@dschneider-pivotal Please let me know if there's any chance we can merge your changes into this PR. There's a few weeks of work in this branch and the sooner we can merge it to develop, the sooner we can stop hitting conflicts.

Reply:
I recommend merging this pull request and then I can redo the object allocation optimization, if needed, in another pull request.

Reply:
I discussed this with Darrel...